### PR TITLE
[add] support saving all pages of a BI Data Table

### LINKED ISSUE #8: 爬虫优化
## 问题
- 现有爬取程序不能满足修改记录的需求
- 飞书自动化每月 200 次限制🚫

## 方案
- 定时任务

## 涉及范围
- 开源文库
- COSCon 2023 数据库

### LINKED ISSUE #217: 开源社官网若干篇文章页面 404
- https://kaiyuanshe.cn/article/KCC-Dalian-A-private-brainstorming-session-about-open-source-business
- https://kaiyuanshe.cn/article/My-story-with-COSCon-From-Passerby-A-to-the-Light-of-the-Community

### PR DESCRIPTION
**Pull request by wangrunlin**
1. [add] support saving all pages of a BI Data Table
    - fix #8
    - fix kaiyuanshe/kaiyuanshe.github.io#217

### CODE REVIEW

**Comment by wangrunlin** [comment]
目前请求数据还没有跑通
```ts
{ code: 91402, msg: 'NOTEXIST', data: {} }
```

File: src/controller/Crawler.ts [hunk 1]
```diff
@@ -20,22 +21,37 @@ import {
...
+                            path = `file/${item.name}`;
+                        const URI = `${OWSBlobHost}/$web/${path}`;
+
+                        await uploadToAzureBlob(file, path, item.type);
+
+                        console.log(`[upload] ${URI}`);
+
+                        yield URI;
+                    }
+    }
```
Comment (TechQuery) [inline src/controller/Crawler.ts]: 老代码先不移动，减少差异对比。

File: src/controller/Crawler.ts [hunk 2]
```diff
@@ -57,36 +73,56 @@ export class CrawlerController {
...
+        const records = await table.getAll(),
+            list: LarkBaseTableModel[] = [];
 
-                        yield URI;
-                    }
+        for (const fields of records) {
+            list.push({
+                id: fields['id'] as string
+            });
+            this.savePage({ source: fields['link'] + '', rootSelector });
```
Comment (TechQuery) [inline src/controller/Crawler.ts]:
```suggestion
            await this.savePage({ source: fields['link'] + '', rootSelector });
```

File: src/model/Crawler.ts [hunk 3]
```diff
@@ -1,39 +1,40 @@
...
     @IsString()
     record: string;
 }
 
-export class LarkBaseTableRecordFileModel {
-    @IsString({ each: true })
-    files: string[];
+export class LarkBaseTableModel {
+    @IsString()
+    id?: string;
```
Comment (TechQuery) [inline src/model/Crawler.ts]:
```suggestion
    id: string;
```

**Review by TechQuery — CHANGES_REQUESTED** [review]

**Comment by wangrunlin** [comment]
@CodiumAI-Agent /review

**Comment by wangrunlin** [comment]
@CodiumAI-Agent /describe

File: src/controller/Crawler.ts (hunk 1 above)
Comment (wangrunlin) [inline src/controller/Crawler.ts]: @CodiumAI-Agent /ask What's this?

**Comment by wangrunlin** [comment]
@TechQuery mobx-lark 升级到 2.0.0-rc.3 后，运行 `pnpm dev` 会出现以下报错
```
Error [ERR_REQUIRE_ESM]: require() of ES Module /workspace/KYS-service/node_modules/.pnpm/native-file-system-adapter@3.0.1/node_modules/native-file-system-adapter/src/es6.js from /workspace/KYS-service/node_modules/.pnpm/mobx-restful@1.0.1_jsdom@22.1.0_mobx@5.15.7_typescript@5.3.3/node_modules/mobx-restful/dist/index.js not supported.
Instead change the require of es6.js in /workspace/KYS-service/node_modules/.pnpm/mobx-restful@1.0.1_jsdom@22.1.0_mobx@5.15.7_typescript@5.3.3/node_modules/mobx-restful/dist/index.js to a dynamic import() which is available in all CommonJS modules.
    at require.extensions..jsx.require.extensions..js (/tmp/ts-node-dev-hook-2701919942264728.js:114:20)
    at Object.nodeDevHook [as .js] (/workspace/KYS-service/node_modules/.pnpm/ts-node-dev@2.0.0_@types+node@18.19.3_typescript@5.3.3/node_modules/ts-node-dev/lib/hook.js:63:13)
    at Object.<anonymous> (/workspace/KYS-service/node_modules/.pnpm/mobx-restful@1.0.1_jsdom@22.1.0_mobx@5.15.7_typescript@5.3.3/node_modules/mobx-restful/dist/index.js:1:1048)
    at Module._compile (/workspace/KYS-service/node_modules/.pnpm/source-map-support@0.5.21/node_modules/source-map-support/source-map-support.js:568:25)
[ERROR] 09:58:34 Error: require() of ES Module /workspace/KYS-service/node_modules/.pnpm/native-file-system-adapter@3.0.1/node_modules/native-file-system-adapter/src/es6.js from /workspace/KYS-service/node_modules/.pnpm/mobx-restful@1.0.1_jsdom@22.1.0_mobx@5.15.7_typescript@5.3.3/node_modules/mobx-restful/dist/index.js not supported.
Instead change the require of es6.js in /workspace/KYS-service/node_modules/.pnpm/mobx-restful@1.0.1_jsdom@22.1.0_mobx@5.15.7_typescript@5.3.3/node_modules/mobx-restful/dist/index.js to a dynamic import() which is available in all CommonJS modules.
```

目前计划更新到弃用的版本 `1.1.1`

**Comment by TechQuery** [comment]
> @TechQuery mobx-lark 升级到 2.0.0-rc.3 后，运行 `pnpm dev` 会出现以下报错
> 
> ```
> Error [ERR_REQUIRE_ESM]: require() of ES Module /workspace/KYS-service/node_modules/.pnpm/native-file-system-adapter@3.0.1/node_modules/native-file-system-adapter/src/es6.js from /workspace/KYS-service/node_modules/.pnpm/mobx-restful@1.0.1_jsdom@22.1.0_mobx@5.15.7_typescript@5.3.3/node_modules/mobx-restful/dist/index.js not supported.
> Instead change the require of es6.js in /workspace/KYS-service/node_modules/.pnpm/mobx-restful@1.0.1_jsdom@22.1.0_mobx@5.15.7_typescript@5.3.3/node_modules/mobx-restful/dist/index.js to a dynamic import() which is available in all CommonJS modules.
>     at require.extensions..jsx.require.extensions..js (/tmp/ts-node-dev-hook-2701919942264728.js:114:20)
>     at Object.nodeDevHook [as .js] (/workspace/KYS-service/node_modules/.pnpm/ts-node-dev@2.0.0_@types+node@18.19.3_typescript@5.3.3/node_modules/ts-node-dev/lib/hook.js:63:13)
>     at Object.<anonymous> (/workspace/KYS-service/node_modules/.pnpm/mobx-restful@1.0.1_jsdom@22.1.0_mobx@5.15.7_typescript@5.3.3/node_modules/mobx-restful/dist/index.js:1:1048)
>     at Module._compile (/workspace/KYS-service/node_modules/.pnpm/source-map-support@0.5.21/node_modules/source-map-support/source-map-support.js:568:25)
> [ERROR] 09:58:34 Error: require() of ES Module /workspace/KYS-service/node_modules/.pnpm/native-file-system-adapter@3.0.1/node_modules/native-file-system-adapter/src/es6.js from /workspace/KYS-service/node_modules/.pnpm/mobx-restful@1.0.1_jsdom@22.1.0_mobx@5.15.7_typescript@5.3.3/node_modules/mobx-restful/dist/index.js not supported.
> Instead change the require of es6.js in /workspace/KYS-service/node_modules/.pnpm/mobx-restful@1.0.1_jsdom@22.1.0_mobx@5.15.7_typescript@5.3.3/node_modules/mobx-restful/dist/index.js to a dynamic import() which is available in all CommonJS modules.
> ```
> 
> 目前计划更新到弃用的版本 `1.1.1`

该包的 ReadMe 有写解决方案，之前开会也讲过。

**Comment by wangrunlin** [comment]
@TechQuery 目前主分支代码流程跑不通，疑似 `web-fetch` 的 `loadPage` 函数有问题，我单独建立一个测试文件没有，但是在该服务中集成无法获取到 `document`。

image

**Comment by TechQuery** [comment]
> @TechQuery 目前主分支代码流程跑不通，疑似 `web-fetch` 的 `loadPage` 函数有问题，我单独建立一个测试文件没有，但是在该服务中集成无法获取到 `document`。
> 
> image

`try {} catch {}` 这个函数看看

File: src/controller/Crawler.ts [hunk 4]
```diff
@@ -81,12 +69,38 @@ export class CrawlerController {
...
+            rootSelector,
+            baseURI
+        })) {
+            await uploadToAzureBlob(data, `${folder}/${name}`, MIME);
+
+            console.log(`[upload] ${baseURI}${name}`);
+        }
+
+        return { scope, baseURI };
+    }
```
Comment (TechQuery) [inline src/controller/Crawler.ts]: 和上面的方法位置换一下，看看能不能减少变更差异对比。

**Review by TechQuery — CHANGES_REQUESTED** [review]

**Comment by wangrunlin** [comment]
> > @TechQuery 目前主分支代码流程跑不通，疑似 `web-fetch` 的 `loadPage` 函数有问题，我单独建立一个测试文件没有，但是在该服务中集成无法获取到 `document`。
> > image
> 
> `try {} catch {}` 这个函数看看

```
  <-- POST /crawler/task/page
TimeoutError: Timed out after waiting 30000ms
    at /Users/leo/code/kaiyuanshe/KYS-service/node_modules/.pnpm/puppeteer-core@22.15.0/node_modules/puppeteer-core/src/common/util.ts:322:17
    at /Users/leo/code/kaiyuanshe/KYS-service/node_modules/.pnpm/puppeteer-core@22.15.0/node_modules/puppeteer-core/lib/cjs/third_party/rxjs/rxjs.js:3986:35
    at OperatorSubscriber2._this._next (/Users/leo/code/kaiyuanshe/KYS-service/node_modules/.pnpm/puppeteer-core@22.15.0/node_modules/puppeteer-core/lib/cjs/third_party/rxjs/rxjs.js:1055:13)
    at OperatorSubscriber2.Subscriber2.next (/Users/leo/code/kaiyuanshe/KYS-service/node_modules/.pnpm/puppeteer-core@22.15.0/node_modules/puppeteer-core/lib/cjs/third_party/rxjs/rxjs.js:678:16)
    at AsyncAction2.<anonymous> (/Users/leo/code/kaiyuanshe/KYS-service/node_modules/.pnpm/puppeteer-core@22.15.0/node_modules/puppeteer-core/lib/cjs/third_party/rxjs/rxjs.js:4863:24)
    at AsyncAction2._execute (/Users/leo/code/kaiyuanshe/KYS-service/node_modules/.pnpm/puppeteer-core@22.15.0/node_modules/puppeteer-core/lib/cjs/third_party/rxjs/rxjs.js:1974:16)
    at AsyncAction2.execute (/Users/leo/code/kaiyuanshe/KYS-service/node_modules/.pnpm/puppeteer-core@22.15.0/node_modules/puppeteer-core/lib/cjs/third_party/rxjs/rxjs.js:1963:26)
    at AsyncScheduler2.flush (/Users/leo/code/kaiyuanshe/KYS-service/node_modules/.pnpm/puppeteer-core@22.15.0/node_modules/puppeteer-core/lib/cjs/third_party/rxjs/rxjs.js:2235:30)
    at listOnTimeout (node:internal/timers:573:17)
    at processTimers (node:internal/timers:514:7) {
  [cause]: undefined
}
  --> POST /crawler/task/page 500 31s 851b
```
image

**Comment by TechQuery** [comment]
> ```
>   <-- POST /crawler/task/page
> TimeoutError: Timed out after waiting 30000ms
>     at /Users/leo/code/kaiyuanshe/KYS-service/node_modules/.pnpm/puppeteer-core@22.15.0/node_modules/puppeteer-core/src/common/util.ts:322:17
>     at /Users/leo/code/kaiyuanshe/KYS-service/node_modules/.pnpm/puppeteer-core@22.15.0/node_modules/puppeteer-core/lib/cjs/third_party/rxjs/rxjs.js:3986:35
>     at OperatorSubscriber2._this._next (/Users/leo/code/kaiyuanshe/KYS-service/node_modules/.pnpm/puppeteer-core@22.15.0/node_modules/puppeteer-core/lib/cjs/third_party/rxjs/rxjs.js:1055:13)
>     at OperatorSubscriber2.Subscriber2.next (/Users/leo/code/kaiyuanshe/KYS-service/node_modules/.pnpm/puppeteer-core@22.15.0/node_modules/puppeteer-core/lib/cjs/third_party/rxjs/rxjs.js:678:16)
>     at AsyncAction2.<anonymous> (/Users/leo/code/kaiyuanshe/KYS-service/node_modules/.pnpm/puppeteer-core@22.15.0/node_modules/puppeteer-core/lib/cjs/third_party/rxjs/rxjs.js:4863:24)
>     at AsyncAction2._execute (/Users/leo/code/kaiyuanshe/KYS-service/node_modules/.pnpm/puppeteer-core@22.15.0/node_modules/puppeteer-core/lib/cjs/third_party/rxjs/rxjs.js:1974:16)
>     at AsyncAction2.execute (/Users/leo/code/kaiyuanshe/KYS-service/node_modules/.pnpm/puppeteer-core@22.15.0/node_modules/puppeteer-core/lib/cjs/third_party/rxjs/rxjs.js:1963:26)
>     at AsyncScheduler2.flush (/Users/leo/code/kaiyuanshe/KYS-service/node_modules/.pnpm/puppeteer-core@22.15.0/node_modules/puppeteer-core/lib/cjs/third_party/rxjs/rxjs.js:2235:30)
>     at listOnTimeout (node:internal/timers:573:17)
>     at processTimers (node:internal/timers:514:7) {
>   [cause]: undefined
> }
>   --> POST /crawler/task/page 500 31s 851b
> ```
> 
> image

经查是较新版 Puppteer core 无法正常启动本地浏览器，但升级到最新版 bug 依旧：https://github.com/puppeteer/puppeteer/issues/13012

**Review by TechQuery — APPROVED** [review]

**Pull request merged**